File: backend/src/utils/pdfGenerator.js
const PDFDocument = require('pdfkit');
const fs = require('fs');
const path = require('path');

class PDFGenerator {
  async generateInvoicePDF(invoice, client, lineItems) {
    try {
      const doc = new PDFDocument({ margin: 50 });
      const filename = `${invoice.invoice_number}.pdf`;
      const uploadDir = process.env.UPLOAD_DIR || './uploads';
      const filepath = path.join(uploadDir, 'invoices', filename);

      // Ensure directory exists
      await fs.promises.mkdir(path.dirname(filepath), { recursive: true });

      // Pipe PDF to file
      doc.pipe(fs.createWriteStream(filepath));

      // Header
      this.addHeader(doc, invoice, client);
      
      // Invoice details
      this.addInvoiceDetails(doc, invoice, client);
      
      // Line items table
      this.addLineItemsTable(doc, lineItems);
      
      // Totals
      this.addTotals(doc, invoice);
      
      // Footer
      this.addFooter(doc);

      // Finalize PDF
      doc.end();

      return new Promise((resolve, reject) => {
        doc.on('end', () => {
          resolve(filepath);
        });
        doc.on('error', reject);
      });
    } catch (error) {
      throw new Error(`PDF generation failed: ${error.message}`);
    }
  }

  addHeader(doc, invoice, client) {
    // Company logo and info
    doc.fontSize(20)
       .text('Tej IT Solutions', 50, 50)
       .fontSize(10)
       .text('AWS Client Billing & Management', 50, 75)
       .text('Email: billing@tejit.com', 50, 90);

    // Invoice title
    doc.fontSize(24)
       .text('INVOICE', 400, 50, { align: 'right' })
       .fontSize(12)
       .text(`Invoice #: ${invoice.invoice_number}`, 400, 80, { align: 'right' })
       .text(`Date: ${new Date(invoice.invoice_date).toLocaleDateString()}`, 400, 95, { align: 'right' });

    // Line separator
    doc.moveTo(50, 140)
       .lineTo(550, 140)
       .stroke();
  }

  addInvoiceDetails(doc, invoice, client) {
    let yPosition = 160;

    // Bill to section
    doc.fontSize(14)
       .text('Bill To:', 50, yPosition)
       .fontSize(12)
       .text(client.client_name, 50, yPosition + 20)
       .text(client.contact_person || '', 50, yPosition + 35)
       .text(client.email, 50, yPosition + 50);

    if (client.billing_address) {
      const addressLines = client.billing_address.split('\n');
      addressLines.forEach((line, index) => {
        doc.text(line, 50, yPosition + 65 + (index * 15));
      });
    }

    // GST information
    if (client.gst_registered && client.gst_number) {
      doc.text(`GST Number: ${client.gst_number}`, 50, yPosition + 120);
    }
  }

  addLineItemsTable(doc, lineItems) {
    const tableTop = 280;
    const itemCodeX = 50;
    const descriptionX = 150;
    const quantityX = 350;
    const rateX = 400;
    const amountX = 480;

    // Table header
    doc.fontSize(12)
       .text('Item', itemCodeX, tableTop)
       .text('Description', descriptionX, tableTop)
       .text('Qty', quantityX, tableTop)
       .text('Rate', rateX, tableTop)
       .text('Amount', amountX, tableTop);

    // Header line
    doc.moveTo(itemCodeX, tableTop + 15)
       .lineTo(550, tableTop + 15)
       .stroke();

    let yPosition = tableTop + 30;

    // Line items
    lineItems.forEach((item, index) => {
      const amount = parseFloat(item.quantity) * parseFloat(item.rate) * (1 - parseFloat(item.discount || 0) / 100);
      
      doc.fontSize(10)
         .text((index + 1).toString(), itemCodeX, yPosition)
         .text(item.description, descriptionX, yPosition, { width: 180 })
         .text(item.quantity.toString(), quantityX, yPosition)
         .text(`${item.currency} ${parseFloat(item.rate).toFixed(4)}`, rateX, yPosition)
         .text(`${item.currency} ${amount.toFixed(2)}`, amountX, yPosition);

      yPosition += 25;
    });

    // Bottom line
    doc.moveTo(itemCodeX, yPosition)
       .lineTo(550, yPosition)
       .stroke();
  }

  addTotals(doc, invoice) {
    const totalsX = 400;
    let yPosition = 500;

    doc.fontSize(12)
       .text('Subtotal:', totalsX, yPosition)
       .text(`${invoice.currency || 'USD'} ${invoice.subtotal || 0}`, totalsX + 100, yPosition, { align: 'right' });

    if (invoice.gst_amount > 0) {
      yPosition += 20;
      doc.text('GST (18%):', totalsX, yPosition)
         .text(`${invoice.currency || 'USD'} ${invoice.gst_amount}`, totalsX + 100, yPosition, { align: 'right' });
    }

    yPosition += 20;
    doc.fontSize(14)
       .text('Total:', totalsX, yPosition)
       .text(`${invoice.currency || 'USD'} ${invoice.total_amount || 0}`, totalsX + 100, yPosition, { align: 'right' });
  }

  addFooter(doc) {
    const footerY = 700;
    
    doc.fontSize(10)
       .text('Payment Terms: Net 30 days', 50, footerY)
       .text('Thank you for your business!', 50, footerY + 15);

    doc.text('Tej IT Solutions - AWS Client Billing & Management System', 50, footerY + 60, {
      align: 'center',
      width: 500
    });
  }
}

module.exports = new PDFGenerator();

File: backend/app/tasks/pdf_generator.py
from celery import Celery
from app import create_app, db
from app.models import Invoice
from weasyprint import HTML, CSS
from jinja2 import Template
import os
from datetime import datetime

# Create Celery instance
celery = Celery('pdf_generator')

@celery.task(bind=True)
def generate_invoice_pdf(self, invoice_id):
    """Generate PDF for an invoice"""
    app = create_app()
    
    with app.app_context():
        try:
            invoice = Invoice.query.get(invoice_id)
            if not invoice:
                raise Exception(f"Invoice {invoice_id} not found")
            
            # Create PDF directory if it doesn't exist
            pdf_dir = os.path.join(app.config['UPLOAD_DIR'], 'invoices', 'pdfs')
            os.makedirs(pdf_dir, exist_ok=True)
            
            # Generate PDF filename
            pdf_filename = f"{invoice.invoice_number}.pdf"
            pdf_path = os.path.join(pdf_dir, pdf_filename)
            
            # HTML template for invoice
            html_template = """
            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="utf-8">
                <title>Invoice {{ invoice.invoice_number }}</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 0; padding: 20px; }
                    .header { text-align: center; margin-bottom: 30px; }
                    .company-name { font-size: 24px; font-weight: bold; color: #2563eb; }
                    .invoice-title { font-size: 20px; margin: 20px 0; }
                    .invoice-details { display: flex; justify-content: space-between; margin-bottom: 30px; }
                    .client-info, .invoice-info { width: 45%; }
                    .table { width: 100%; border-collapse: collapse; margin-bottom: 20px; }
                    .table th, .table td { border: 1px solid #ddd; padding: 8px; text-align: left; }
                    .table th { background-color: #f8f9fa; }
                    .totals { text-align: right; margin-top: 20px; }
                    .total-row { font-weight: bold; font-size: 16px; }
                    .footer { margin-top: 40px; text-align: center; font-size: 12px; color: #666; }
                </style>
            </head>
            <body>
                <div class="header">
                    <div class="company-name">TejIT Solutions</div>
                    <div>AWS Cloud Services Provider</div>
                </div>
                
                <div class="invoice-title">INVOICE</div>
                
                <div class="invoice-details">
                    <div class="client-info">
                        <h3>Bill To:</h3>
                        <strong>{{ invoice.client.client_name }}</strong><br>
                        {{ invoice.client.contact_person }}<br>
                        {{ invoice.client.email }}<br>
                        {% if invoice.client.phone %}{{ invoice.client.phone }}<br>{% endif %}
                        {% if invoice.client.billing_address %}
                            {{ invoice.client.billing_address | replace('\n', '<br>') | safe }}
                        {% endif %}
                        {% if invoice.client.gst_number %}
                            <br><strong>GST Number:</strong> {{ invoice.client.gst_number }}
                        {% endif %}
                    </div>
                    
                    <div class="invoice-info">
                        <strong>Invoice Number:</strong> {{ invoice.invoice_number }}<br>
                        <strong>Invoice Date:</strong> {{ invoice.invoice_date.strftime('%B %d, %Y') }}<br>
                        <strong>Due Date:</strong> {{ invoice.due_date.strftime('%B %d, %Y') }}<br>
                        {% if invoice.usd_to_inr_rate %}
                            <strong>USD to INR Rate:</strong> {{ invoice.usd_to_inr_rate }}
                        {% endif %}
                    </div>
                </div>
                
                <table class="table">
                    <thead>
                        <tr>
                            <th>Description</th>
                            <th>Quantity</th>
                            <th>Rate</th>
                            <th>Discount</th>
                            <th>Amount</th>
                        </tr>
                    </thead>
                    <tbody>
                        {% for item in invoice.line_items %}
                        <tr>
                            <td>{{ item.description }}</td>
                            <td>{{ item.quantity }}</td>
                            <td>{{ item.currency.value }} {{ "%.4f"|format(item.rate) }}</td>
                            <td>{{ item.discount }}%</td>
                            <td>{{ item.currency.value }} {{ "%.2f"|format(item.calculate_amount()) }}</td>
                        </tr>
                        {% endfor %}
                    </tbody>
                </table>
                
                <div class="totals">
                    <div>Subtotal: {{ invoice.line_items[0].currency.value if invoice.line_items else 'USD' }} {{ "%.2f"|format(totals.subtotal) }}</div>
                    {% if invoice.gst_applicable %}
                        <div>GST (18%): {{ invoice.line_items[0].currency.value if invoice.line_items else 'USD' }} {{ "%.2f"|format(totals.gst_amount) }}</div>
                    {% endif %}
                    <div class="total-row">Total: {{ invoice.line_items[0].currency.value if invoice.line_items else 'USD' }} {{ "%.2f"|format(totals.total) }}</div>
                </div>
                
                {% if invoice.invoice_notes %}
                <div style="margin-top: 30px;">
                    <strong>Notes:</strong><br>
                    {{ invoice.invoice_notes | replace('\n', '<br>') | safe }}
                </div>
                {% endif %}
                
                <div class="footer">
                    <p>Thank you for your business!</p>
                    <p>This is a computer-generated invoice.</p>
                </div>
            </body>
            </html>
            """
            
            # Render template
            template = Template(html_template)
            totals = invoice.calculate_totals()
            html_content = template.render(invoice=invoice, totals=totals)
            
            # Generate PDF
            HTML(string=html_content).write_pdf(pdf_path)
            
            # Update invoice with PDF path
            invoice.pdf_path = pdf_path
            db.session.commit()
            
            return {
                'status': 'success',
                'message': f'PDF generated successfully for invoice {invoice.invoice_number}',
                'pdf_path': pdf_path
            }
            
        except Exception as e:
            db.session.rollback()
            return {
                'status': 'error',
                'message': f'Failed to generate PDF: {str(e)}'
            }

@celery.task(bind=True)
def generate_bulk_pdfs(self, invoice_ids):
    """Generate PDFs for multiple invoices"""
    app = create_app()
    
    with app.app_context():
        results = []
        
        for invoice_id in invoice_ids:
            try:
                result = generate_invoice_pdf.delay(invoice_id)
                results.append({
                    'invoice_id': invoice_id,
                    'task_id': result.id,
                    'status': 'queued'
                })
            except Exception as e:
                results.append({
                    'invoice_id': invoice_id,
                    'status': 'error',
                    'message': str(e)
                })
        
        return {
            'status': 'success',
            'message': f'Queued PDF generation for {len(invoice_ids)} invoices',
            'results': results
        }